daml-script/runner/src/main/scala/com/daml/script/Runner.scala

```diff
@@ -0,0 +1,341 @@
+import com.typesafe.scalalogging.StrictLogging
+import java.time.Instant
+import java.util
+import java.util.UUID
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scalaz.syntax.tag._
+
+import com.digitalasset.api.util.TimestampConversion.fromInstant
+import com.digitalasset.daml.lf.PureCompiledPackages
+import com.digitalasset.daml.lf.archive.Dar
+import com.digitalasset.daml.lf.data.FrontStack
+import com.digitalasset.daml.lf.data.Ref._
+import com.digitalasset.daml.lf.language.Ast._
+import com.digitalasset.daml.lf.speedy.{Compiler, Pretty, SExpr, Speedy, SValue}
+import com.digitalasset.daml.lf.speedy.SBuiltin._
+import com.digitalasset.daml.lf.speedy.SExpr._
+import com.digitalasset.daml.lf.speedy.SResult._
+import com.digitalasset.daml.lf.speedy.SValue._
+import com.digitalasset.daml.lf.value.Value.{AbsoluteContractId, RelativeContractId}
+import com.digitalasset.ledger.api.domain.LedgerId
+import com.digitalasset.ledger.api.refinements.ApiTypes.ApplicationId
+import com.digitalasset.ledger.api.v1.command_service.SubmitAndWaitRequest
+import com.digitalasset.ledger.api.v1.commands._
+import com.digitalasset.ledger.api.v1.event.{CreatedEvent}
+import com.digitalasset.ledger.api.v1.transaction.TreeEvent
+import com.digitalasset.ledger.api.v1.transaction_filter.{
+  Filters,
+  TransactionFilter,
+  InclusiveFilters
+}
+import com.digitalasset.ledger.api.v1.value.{Identifier => ApiIdentifier}
+import com.digitalasset.ledger.api.validation.ValueValidator
+import com.digitalasset.ledger.client.LedgerClient
+import com.digitalasset.platform.participant.util.LfEngineToApi.{
+  toApiIdentifier,
+  lfValueToApiRecord,
+  lfValueToApiValue
+}
+
+class Runner(dar: Dar[(PackageId, Package)], applicationId: ApplicationId) extends StrictLogging {
+
+  val darMap: Map[PackageId, Package] = dar.all.toMap
+  val compiler = Compiler(darMap)
+  val scriptModuleName = DottedName.assertFromString("Daml.Script")
+  val scriptPackageId: PackageId = dar.all
+    .find {
+      case (pkgId, pkg) => pkg.modules.contains(scriptModuleName)
+    }
+    .get
+    ._1
+  val stdlibPackageId =
+    dar.all
+      .find {
+        case (pkgId, pkg) =>
+          pkg.modules.contains(DottedName.assertFromString("DA.Internal.LF"))
+      }
+      .get
+      ._1
+
+  // We overwrite the definition of toLedgerValue with an identity function.
+  // This is a type error but Speedy doesn’t care about the types and the only thing we do
+  // with the result is convert it to ledger values/record so this is safe.
+  val definitionMap =
+    compiler.compilePackages(darMap.keys) +
+      (LfDefRef(
+        Identifier(
+          scriptPackageId,
+          QualifiedName(scriptModuleName, DottedName.assertFromString("fromLedgerValue")))) ->
+        SEMakeClo(Array(), 1, SEVar(1)))
+  val compiledPackages = PureCompiledPackages(darMap, definitionMap).right.get
+
+  def toLedgerRecord(v: SValue) = {
```

Comment:
Could some of this be shared between triggers and daml-script?

Reply:
We could probably share a bit, e.g., the logic for converting `AnyTemplate` but most of the conversion are specific to the types used in the trigger/script library so the shared code isn’t as large as one might hope. I’ll first cleanup the conversion logic here and then see what we can factor out.